### PR TITLE
feature #532: Add endpoint for group member on schedule

### LINKED ISSUE #532: Add endpoint for group member on schedule
Create an endpoint, that is used for loggined users. If user is loggined:
- check if he/she want to get info about his/her group. if it's his group, return all week events, otherwise look general for group endpoint;
- create query `showOwnSelective`, if it's true you need to filter all selective disciplines and return only selected by user, otherwise return all selective;
- return not only lessons (events that have lesson entity), but other too (events that don't have lesson entity);

Create swagger documentation for this endpoint and response type;

### PR DESCRIPTION
**Pull request by kropli**
- implemented getGroupEvents method
- added getAllGroupEvents methods in ScheduleService
- added GroupEventResponse
- added integration tests for getGroupEvents
- described swagger documentation

closes #532

### CODE REVIEW

File: src/v2/api/services/UserService.ts [hunk 1]
```diff
@@ -544,4 +547,36 @@ export class UserService {
...
+    const user = await this.getUser(userId);
+    const isUserGroup = user.group.id === groupId;
+    let groupEvents: DbEvent[];
+    if (isUserGroup) {
+      groupEvents = await this.scheduleService.getAllGroupEvents(groupId, null);
+    } else {
+      groupEvents = await this.scheduleService.getGeneralGroupEvents(groupId, null);
+    }
+
+    if (isOwnSelected) {
```
Comment (just-hoshinon) [inline src/v2/api/services/UserService.ts:565]: if it's not user's group, you don't need to filter selective disciplines

File: src/v2/api/services/UserService.ts [hunk 2]
```diff
@@ -544,4 +547,36 @@ export class UserService {
...
+      const userSelective = await this.getSelectiveDisciplines(userId);
+      for (const event of groupEvents) {
+        event.lessons = event.lessons
```
Comment (just-hoshinon) [inline src/v2/api/services/UserService.ts:568]: no filtration of events

**Review by just-hoshinon — CHANGES_REQUESTED** [review]

File: src/v2/api/services/ScheduleService.ts [hunk 3]
```diff
@@ -237,4 +239,55 @@ export class ScheduleService {
...
+      return await this.getGeneralGroupEvents(groupId, null);
+    }
+
+    if (isOwnSelected) {
+      const userSelective = await this.userService.getSelectiveDisciplines(userId);
+      await filterAsync(groupEvents,  async (event) => {
+        const lesson = event.lessons[0];
+        if ((await this.disciplineRepository.findById(lesson.disciplineType.disciplineId)).isSelective) {
+          return userSelective.some((selectiveDiscipline) =>
+            selectiveDiscipline.id === lesson.disciplineType.disciplineId);
```
Comment (just-hoshinon) [inline src/v2/api/services/ScheduleService.ts:284]: you can use ArrayUtils method:
```ts
some(userSelective, 'id', lesson.disciplineType.disciplineId);
```

File: src/v2/api/services/ScheduleService.ts [hunk 4]
```diff
@@ -237,4 +239,55 @@ export class ScheduleService {
...
+      where: {
+        groupId: id,
+        endTime: {
+          gte: startOfWeek,
+        },
+        startTime: {
+          lte: endOfWeek,
+        },
+      },
+    }) as unknown as DbEvent[];
```
Comment (just-hoshinon) [inline src/v2/api/services/ScheduleService.ts:255]: It's better to declare type in EventRepository:
```ts
async findMany(...): Promise<DbEvent[]> {

}
```

File: src/v2/api/services/ScheduleService.ts [hunk 5]
```diff
@@ -237,4 +239,55 @@ export class ScheduleService {
       discipline: result.discipline,
     };
   }
+
+  async getAllGroupEvents (id, week) {
```
Comment (just-hoshinon) [inline src/v2/api/services/ScheduleService.ts:243]: you can rename id to groupId

File: src/v2/api/services/ScheduleService.ts [hunk 6]
```diff
@@ -237,4 +239,55 @@ export class ScheduleService {
...
+    const { startOfWeek, endOfWeek } = week ? await this.dateService.getDatesOfWeek(week) : this.dateService.getDatesOfCurrentWeek();
+    const events = await this.eventRepository.findMany({
+      where: {
+        groupId: id,
```
Comment (just-hoshinon) [inline src/v2/api/services/ScheduleService.ts:247]: after changing id to groupId you can write just `groupId,`

File: src/v2/api/services/ScheduleService.ts [hunk 7]
```diff
@@ -237,4 +239,55 @@ export class ScheduleService {
...
+
+    if (isOwnSelected) {
+      const userSelective = await this.userService.getSelectiveDisciplines(userId);
+      await filterAsync(groupEvents,  async (event) => {
+        const lesson = event.lessons[0];
+        if ((await this.disciplineRepository.findById(lesson.disciplineType.disciplineId)).isSelective) {
+          return userSelective.some((selectiveDiscipline) =>
+            selectiveDiscipline.id === lesson.disciplineType.disciplineId);
+        } else {
+          return true;
```
Comment (just-hoshinon) [inline src/v2/api/services/ScheduleService.ts:286]: making else in a such way is a bad practice. you can revert your if-statement, making it more readable:
```ts
const discipline = ...;
if (!discipline.isSelective) return true;

return some(...);
```

File: src/v2/api/responses/AllEventResponse.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  @ApiProperty()
+    name: string;
+
+  @ApiProperty()
+    startTime: Date;
+
+  @ApiProperty()
+    endTime: Date;
+
+  @ApiPropertyOptional({
```
Comment (just-hoshinon) [inline src/v2/api/responses/AllEventResponse.ts:17]: It's not optional. Optional means undefined. This field can be type or null, not undefined

File: src/v2/api/controllers/ScheduleController.ts [hunk 9]
```diff
@@ -180,4 +182,32 @@ export class ScheduleController {
...
+    type: Boolean,
+    name: 'showOwnSelective',
+  })
+  async getGroupEvents (
+      @Request() req,
+      @Param('groupId', GroupByIdPipe) groupId: string,
+      @Query('showOwnSelective', ConvertToBooleanPipe) showOwnSelective,
+  ) {
+    const events = await this.scheduleService.getGroupEvents(req.user.id, groupId, showOwnSelective);
+    return this.scheduleMapper.getGeneralEvents(events);
```
Comment (just-hoshinon) [inline src/v2/api/controllers/ScheduleController.ts:211]: it will be better to rename this method to getEvents

**Review by just-hoshinon — CHANGES_REQUESTED** [review]

**Review by just-hoshinon — CHANGES_REQUESTED** [review]
Make endpoint return week and events in object

File: src/v2/mappers/ScheduleMapper.ts [hunk 10]
```diff
@@ -7,13 +7,13 @@ import { EventResponse } from '../api/responses/EventResponse';
 
 @Injectable()
 export class ScheduleMapper {
-  getGeneralEvents (events: DbEvent[]) {
+  getEvents (events: DbEvent[])  {
```
Comment (JessFreak) [inline src/v2/mappers/ScheduleMapper.ts:10]: extra space

**Review by JessFreak — APPROVED** [review]

File: src/v2/api/controllers/ScheduleController.ts [hunk 11]
```diff
@@ -180,4 +188,46 @@ export class ScheduleController {
     const result = await this.scheduleService.createGroupEvent(body);
     return this.scheduleMapper.getEvent(result.event, result.discipline);
   }
+
+  @Access('groups.$groupId.events.get')
```
Comment (PavloVlasiuk) [inline src/v2/api/controllers/ScheduleController.ts:192]: add authorization to swagger documentation

File: src/v2/api/controllers/ScheduleController.ts [hunk 12]
```diff
@@ -180,4 +188,46 @@ export class ScheduleController {
...
+  })
+  async getGroupEvents (
+      @Request() req,
+      @Param('groupId', GroupByIdPipe) groupId: string,
+      @Query('showOwnSelective', ConvertToBooleanPipe) showOwnSelective,
+      @Query('week') week: number,
+  ) {
+    const result = await this.scheduleService.getGroupEvents(req.user.id, groupId, showOwnSelective, week);
+    return {
+      events: this.scheduleMapper.getEvents(result.events),
```
Comment (PavloVlasiuk) [inline src/v2/api/controllers/ScheduleController.ts:229]: describe returned value disciplineType correctly. Follow this example 
image

File: src/v2/api/controllers/ScheduleController.ts [hunk 13]
```diff
@@ -180,4 +188,46 @@ export class ScheduleController {
...
   }
+
+  @Access('groups.$groupId.events.get')
+  @Get('/groups/:groupId/events')
+  @ApiOkResponse({
+    type: EventsResponse,
+  })
+  @ApiBadRequestResponse({
+    description: `\n
+    InvalidEntityIdException:
```
Comment (PavloVlasiuk) [inline src/v2/api/controllers/ScheduleController.ts:199]: this exception must be InvalidGroupIdException

File: src/v2/api/services/ScheduleService.ts [hunk 14]
```diff
@@ -115,6 +133,11 @@ export class ScheduleService {
       week > endWeek || week < startWeek ||
       (event.period === Period.EVERY_FORTNIGHT && startWeek % 2 !== week % 2)
     ) {
+      const res1 = event.period === Period.EVERY_FORTNIGHT;
+      const res2 = startWeek;
+      const res3 = week;
+      const res4 = startWeek % 2 !== week % 2;
+      console.log('res1: ', res1, ' res2: ', res2, ' res3: ', res3, ' res4: ', res4);
```
Comment (JessFreak) [inline src/v2/api/services/ScheduleService.ts:140]: did not clean colnsole.log

**Review by JessFreak — APPROVED** [review]

File: src/v2/api/services/ScheduleService.ts [hunk 15]
```diff
@@ -81,11 +89,24 @@ export class ScheduleService {
...
+
+    const result  = events
       .filter((event) => {
         const indexOfLesson = this.getIndexOfLesson(event.period, event.startTime, endOfWeek);
         return indexOfLesson !== null;
       });
+
+    for (const event of result) {
+      if (event.period !== Period.NO_PERIOD) {
+        await this.setWeekTime(event, week);
```
Comment (f-ogoog) [inline src/v2/api/services/ScheduleService.ts:102]: period in this case can not be NO_PERIOD

**Pull request merged**